verifier/verifier.go

```diff
@@ -0,0 +1,494 @@
+	// load trust policy
+	policyDocument, err := trustpolicy.LoadDocument()
+	if err != nil {
+		return nil, err
+	}
+
+	return NewVerifier(policyDocument, plugin.NewCLIManager(dir.PluginFS()))
+}
+
+// NewVerifier creates a new verifier given trustPolicy and pluginManager
+func NewVerifier(trustPolicy *trustpolicy.Document, pluginManager plugin.Manager) (notation.Verifier, error) {
+	if err := trustPolicy.Validate(); err != nil {
+		return nil, err
+	}
+	return &verifier{
+		trustPolicyDoc: trustPolicy,
+		pluginManager:  pluginManager,
+	}, nil
+}
+
+// Verify verifies the signature blob and returns the verified descriptor
+// upon successful verification.
+func (v *verifier) Verify(ctx context.Context, desc ocispec.Descriptor, signature []byte, opts notation.VerifyOptions) (*notation.VerificationOutcome, error) {
+	artifactRef := opts.ArtifactReference
+	envelopeMediaType := opts.SignatureMediaType
+	pluginConfig := opts.PluginConfig
+
+	trustPolicy, err := v.trustPolicyDoc.GetApplicableTrustPolicy(artifactRef)
+	if err != nil {
+		return nil, notation.ErrorNoApplicableTrustPolicy{Msg: err.Error()}
+	}
+	// ignore the error since we already validated the policy document
+	verificationLevel, _ := trustPolicy.SignatureVerification.GetVerificationLevel()
+
+	outcome := &notation.VerificationOutcome{
+		RawSignature:      signature,
+		VerificationLevel: verificationLevel,
+	}
+	// verificationLevel is skip
+	if reflect.DeepEqual(verificationLevel, trustpolicy.LevelSkip) {
+		return outcome, nil
+	}
+	err = v.processSignature(ctx, signature, envelopeMediaType, trustPolicy, pluginConfig, outcome)
+	if err != nil {
+		outcome.Error = err
+		return outcome, err
+	}
+
+	payload := &envelope.Payload{}
+	err = json.Unmarshal(outcome.EnvelopeContent.Payload.Content, payload)
+	if err != nil {
+		outcome.Error = err
+		return outcome, err
+	}
+
+	if !content.Equal(payload.TargetArtifact, desc) {
+		outcome.Error = errors.New("content descriptor mismatch")
+	}
+	return outcome, outcome.Error
+}
+
+func (v *verifier) processSignature(ctx context.Context, sigBlob []byte, envelopeMediaType string, trustPolicy *trustpolicy.TrustPolicy, pluginConfig map[string]string, outcome *notation.VerificationOutcome) error {
+
+	// verify integrity first. notation will always verify integrity no matter what the signing scheme is
+	envContent, integrityResult := verifyIntegrity(sigBlob, envelopeMediaType, outcome)
+	outcome.EnvelopeContent = envContent
+	outcome.VerificationResults = append(outcome.VerificationResults, integrityResult)
+	if integrityResult.Error != nil {
+		return integrityResult.Error
+	}
+
+	// check if we need to verify using a plugin
+	var pluginCapabilities []proto.Capability
+	verificationPluginName, err := getVerificationPlugin(&outcome.EnvelopeContent.SignerInfo)
+	// use plugin, but getPluginName returns an error
+	if err != nil && err != errExtendedAttributeNotExist {
+		return err
+	}
+	var installedPlugin plugin.Plugin
+	if err == nil {
```

Comment:
The logic will be more readable if 
```suggestion
	if verificationPluginName != "" {
```